### PR TITLE
feat: add SOAP undelete

### PR DESCRIPTION
**Pull request by scolladon**
Add SOAP undelete method

### CODE REVIEW

File: test/soap.test.ts [hunk 1]
```diff
@@ -155,3 +155,78 @@ describe('convert and merge', () => {
...
+  let contactIds: string[];
+
+  beforeAll(async () => {
+    // Arrange
+    const insertedContacts = await conn.sobject('Lead').create(contacts);
+    contactIds = insertedContacts.map((contact) => contact.id!);
+
+    await Promise.all(
+      contactIds.map((id) => conn.sobject('Contact').destroy(id)),
+    );
```
Comment (cristiand391) [inline test/soap.test.ts:186]: nit: `sobject.destroy()` takes an array of IDs, that should be faster for tests

**Review by cristiand391 — COMMENTED** [review]
just gave this a quick glance, is this ready for review?

**Comment by scolladon** [comment]
> just gave this a quick glance, is this ready for review?

Yes totally, it is ready for review.
Let me know if there are some checks I should do, questions I should answer, etc.
CI is not running properly on my fork probably because it requires a specific setup I have not yet done

**Comment by cristiand391** [comment]
> Let me know if there are some checks I should do, questions I should answer, etc.
> CI is not running properly on my fork probably because it requires a specific setup I have not yet done

yep, we disable CI runs on forks, you could try running them locally if you got a hub and the salesforce cli:
https://github.com/jsforce/jsforce/blob/main/DEVELOPING.md

I'll prob merge this branch into a separate one created by so we get all e2e to run anyway.

**Comment by scolladon** [comment]
> > Let me know if there are some checks I should do, questions I should answer, etc.
> > CI is not running properly on my fork probably because it requires a specific setup I have not yet done
> 
> yep, we disable CI runs on forks, you could try running them locally if you got a hub and the salesforce cli: [`main`/DEVELOPING.md](https://github.com/jsforce/jsforce/blob/main/DEVELOPING.md?rgh-link-date=2024-07-22T20%3A18%3A42Z)
> 
> I'll prob merge this branch into a separate one created by so we get all e2e to run anyway.

Please go ahead, I'll look at the result tomorrow

**Comment by cristiand391** [comment]
@scolladon I'd pushed a new branch to get checks run on your PR, SOAP tests are failing ⬆️

**Comment by scolladon** [comment]
> @scolladon I'd pushed a new branch to get checks run on your PR, SOAP tests are failing ⬆️

Which branch ? where can I look at the result ?

**Comment by cristiand391** [comment]
> Which branch ? where can I look at the result ?

the checks for this PR: 
Screenshot 2024-07-30 at 10 52 13 AM

branch: https://github.com/jsforce/jsforce/tree/pr/1533

**Comment by scolladon** [comment]
> > Which branch ? where can I look at the result ?
> 
> the checks for this PR: Screenshot 2024-07-30 at 10 52 13 AM
> 
> branch: [`pr`/1533](https://github.com/jsforce/jsforce/tree/pr/1533?rgh-link-date=2024-07-30T13%3A52%3A51Z)

Thanks @cristiand391 I had a look at the issue but as I'm not able to test locally neither in my CI I don't know really how to reproduce and so I do not know how to fix it.
I'll try a commit to implement the undelete as the other CUD operation are done.

**Comment by scolladon** [comment]
Hi @cristiand391 

Does the last commit fixes the issue int the test ?
Is there a documentation to learn how to test jsforce locally ?

**Comment by scolladon** [comment]
Hi @cristiand391

I managed to test locally and now it feels better
Any help here to make it move forward ?
I really think having the undelete in jsforce could be an asset for any tooling (sfdx plugins per exemple) using the lib

File: src/api/soap/schema.ts [hunk 2]
```diff
@@ -489,7 +489,7 @@ export const ApiSchemas = {
   UndeleteResult: {
     type: 'UndeleteResult',
     props: {
-      errors: ['Error'],
+      errors: ['?', 'Error'],
```
Comment (cristiand391) [inline src/api/soap/schema.ts:492]: did you update this manually or via an updated WSDL? See the comment at the top of this file

File: src/api/soap.ts [hunk 3]
```diff
@@ -240,6 +243,24 @@ export class SoapApi<S extends Schema> {
...
+  undelete(id: string): Promise<UndeleteResult>;
+  undelete(ids: string | string[]): Promise<UndeleteResult | UndeleteResult[]>;
+  undelete(ids: string | string[]) {
+    const schema = Array.isArray(ids)
+      ? [ApiSchemas.UndeleteResult]
+      : ApiSchemas.UndeleteResult;
+    const args = {
+      '@xmlns': 'urn:partner.soap.sforce.com',
+      '@xmlns:ns1': 'sobject.partner.soap.sforce.com',
+      'ns1:ids': ids,
```
Comment (cristiand391) [inline src/api/soap.ts:260]: have you tried undeleting a single ID (as a string, not a 1-item array)? docs mention an array only:
https://developer.salesforce.com/docs/atlas.en-us.api.meta/api/sforce_api_calls_undelete.htm

File: test/soap.test.ts [hunk 4]
```diff
@@ -155,3 +156,87 @@ describe('convert and merge', () => {
...
+  });
+
+  /**
+   *
+   */
+  it('should undelete deleted contacts', async () => {
+    // Act
+    const res: UndeleteResult[] = (await conn.soap.undelete(
+      contactIds,
+    )) as unknown as UndeleteResult[];
```
Comment (cristiand391) [inline test/soap.test.ts:195]: this type casting can be removed (if you were seeing TS errors locally you need to build jsforce first by running `npm run build:node:cjs`)

**Comment by cristiand391** [comment]
> I managed to test locally and now it feels better
Any help here to make it move forward ?
I really think having the undelete in jsforce could be an asset for any tooling (sfdx plugins per exemple) using the lib

@scolladon sorry, I missed your last comment.

Code looks good, just have a question about the single ID undelete being supported.

File: src/api/soap.ts (hunk 3 above)
Comment (scolladon) [reply src/api/soap.ts:260]: I have not try but you're right.
I'll try and I will fix if necessary

File: src/api/soap/schema.ts (hunk 2 above)
Comment (scolladon) [reply src/api/soap/schema.ts:492]: Manually, I just read the top of the file...
I'll fix it

File: src/api/soap.ts (hunk 3 above)
Comment (scolladon) [reply src/api/soap.ts:255]: As doc mention only array I simplified the code to follow the doc

**Comment by scolladon** [comment]
Hi @cristiand391 

I took your review, I think it is better now, enough for you to have another look and maybe test a bit on your side

**Review by cristiand391 — APPROVED** [review]

**Comment by cristiand391** [comment]
@scolladon thanks!
lgtm 👍🏼 

I pushed a branch with your last commit to get all tests running:
https://github.com/jsforce/jsforce/tree/pr%2F1533

the lint & type check job failed here because it can't find the `UndeleteResult[]` type (run `npm run typecheck` locally to see it):
https://github.com/jsforce/jsforce/actions/runs/11133226626/job/30938839371?pr=1533


You can just remove the type (it's inferred from `soap.undelete`) like this (I would do it but can't push to your branch):
Screenshot 2024-10-01 at 5 52 25 PM

**Comment by scolladon** [comment]
> @scolladon thanks! lgtm 👍🏼
> 
> I pushed a branch with your last commit to get all tests running: [`pr%2F1533`](https://github.com/jsforce/jsforce/tree/pr%2F1533?rgh-link-date=2024-10-01T20%3A55%3A02Z)
> 
> the lint & type check job failed here because it can't find the `UndeleteResult[]` type (run `npm run typecheck` locally to see it): [jsforce/jsforce/actions/runs/11133226626/job/30938839371?pr=1533](https://github.com/jsforce/jsforce/actions/runs/11133226626/job/30938839371?pr=1533)
> 
> You can just remove the type (it's inferred from `soap.undelete`) like this (I would do it but can't push to your branch): Screenshot 2024-10-01 at 5 52 25 PM

Done, thanks for your help

**Review by cristiand391 — APPROVED** [review]

**Comment by cristiand391** [comment]
**QA notes**:

e2e test cover this feature so I just did a manual check to confirm records were being undeleted/failed where expected

✅ throw on invalid IDs
✅ API response matches return type
✅ can undelete records (get server error if it can't)
✅ test clean up records in afterAll hook

**Comment by cristiand391** [comment]
This feature was included in jsforce v3.5.0:
https://github.com/jsforce/jsforce/releases/tag/3.5.0

@scolladon thanks for contributing :)

**Comment by scolladon** [comment]
> This feature was included in jsforce v3.5.0: [`3.5.0` (release)](https://github.com/jsforce/jsforce/releases/tag/3.5.0)
> 
> @scolladon thanks for contributing :)

Thanks for your help and support @cristiand391 
I'm super happy to have contributed

**Pull request merged**